### PR TITLE
Support building Dockerfiles

### PR DESCRIPTION
**Pull request by kklin**

### CODE REVIEW

**Comment by kklin** [comment]
ping @mmussomele

**Comment by mmussomele** [comment]
Sorry got really busy over the last week, I'm halfway through a review

**Comment by kklin** [comment]
Cool, no rush, I just wanted to make sure it was on your radar. It's not blocking anything.

File: minion/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		return
+	}
+
+	var images []db.Image
+	conn.Txn(db.ImageTable).Run(func(view db.Database) error {
+		images = view.SelectFromImage(nil)
+		return nil
+	})
+
+	for _, im := range images {
```
Comment (mmussomele) [inline minion/registry/registry.go]: Should this be done in the original transaction in case things changed between when you read and when you commit?

File: minion/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+func updateRegistry(dk docker.Client, img db.Image) (string, error) {
+	registryImg := "localhost:5000/" + img.Name
+	id, err := dk.Build(registryImg, img.Dockerfile)
+	if err == nil {
+		err = dk.Push("localhost:5000", registryImg)
+	}
+	return id, err
+}
+
+func bootWait(conn db.Conn) {
```
Comment (mmussomele) [inline minion/registry/registry.go]: why does this take a `db.Conn`?

File: minion/scheduler/worker.go [hunk 3]
```diff
@@ -153,7 +153,18 @@ func syncJoinScore(left, right interface{}) int {
...
 	dkc := right.(docker.Container)
 
-	if dbc.Image != dkc.Image || dbc.IP != dkc.IP {
+	compareIDs := dbc.ImageID != ""
+	if compareIDs {
+		if dkc.ImageID != dbc.ImageID {
+			return -1
+		}
+	} else {
+		if dkc.Image != dbc.Image {
```
Comment (mmussomele) [inline minion/scheduler/worker.go]: could this be merged into the else?

Comment (kklin) [inline minion/scheduler/worker.go]: You mean, like
```
if compareIDs {
 	if dkc.ImageID != dbc.ImageID {
 		return -1
 	}
} else if dkc.Image != dbc.Image {
	return -1
{
```
?

I find that pretty hard to follow.

I also considered writing it as
```
compareIDs := dbc.ImageID != ""
namesMatch := dkc.Image == dbc.Image
idsMatch := dkc.ImageID == dbc.ImageID
if (compareIDs && !idsMatch) || (!compareIDs && !namesMatch) {
	return -1
}
```
Wasn't sure if that'd be easier to follow.

File: minion/registry/registry.go (hunk 1 above)
Comment (kklin) [inline minion/registry/registry.go]: Good point

File: minion/registry/registry.go (hunk 2 above)
Comment (kklin) [inline minion/registry/registry.go]: It's not needed anymore, nice catch

**Comment by kklin** [comment]
@mmussomele Updated the PR. I also made some other tweaks:
- Added some documentation to `stitch/README.md`
- Used `util.ToTar` for creating the Dockerfile tarball
- Omitted the Dockerfile from `Container.String` (it's still in `Image`)

I start the `registry` container in the `supervisor`.

File: minion/scheduler/worker.go (hunk 3 above)
Comment (mmussomele) [inline minion/scheduler/worker.go]: Yea I personally like the second one better

Comment (kklin) [inline minion/scheduler/worker.go]: Sounds good to me

**Comment by mmussomele** [comment]
Cool if those are the changes you made then LGTM

File: minion/docker/mock_docker.go [hunk 4]
```diff
@@ -96,6 +111,61 @@ func (dk MockClient) RemoveContainer(opts dkc.RemoveContainerOptions) error {
...
+			return nil, fmt.Errorf("malformed build tarball: %s", err.Error())
+		}
+		if hdr.Name == "Dockerfile" {
+			return ioutil.ReadAll(tarball)
+		}
+	}
+}
+
+// BuildImage builds the requested image.
+func (dk MockClient) BuildImage(opts dkc.BuildImageOptions) error {
```
Comment (ejj) [inline minion/docker/mock_docker.go]: Sigh, wish I had used mockery earlier.

File: stitch/README.md [hunk 5]
```diff
@@ -4,6 +4,38 @@ Documentation for quilt.js
...
+
+#### Dockerfile
+Instead of supplying a link to a pre-built image, Quilt also support building
+images in the cluster. When specifying a `Dockerfile` to be built, an `Image`
+object must be passed to the `Container` constructor.
+
+For example,
+```
+new Container(new Image("my-image-name",
+  "FROM nginx\n" +
```
Comment (ejj) [inline stitch/README.md]: Can you build a Dockerfile in the local filesystem?  I assume you can using read?

I think the docs should show an example of that case -- that's by far going to be the most common.
